File: co/system/backend/app/api/battery.py
"""
Copyright (c) 2026 Clove Twilight
Licensed under the ESAL-1.3 Licence.
See LICENCE.md in the project root for full licence information.
"""

"""
Battery level API endpoints
Devices POST their current battery percentage; clients GET the latest known
value(s). Latest-value-per-device only (no history).
"""

from typing import Dict
from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.models import BatteryLevel, BatteryUpdateResponse
from app.services.battery_service import (
    get_all_levels,
    get_device_level,
    set_device_level,
)
from app.dependencies.battery import verify_battery_access

router = APIRouter(prefix="/api/battery")


@router.post("", response_model=BatteryUpdateResponse)
async def update_battery(
    device: str = Query(..., min_length=1, max_length=64, description="Device name, e.g. 'iphone'"),
    level: int = Query(..., ge=0, le=100, description="Battery percentage, 0-100"),
    authenticated: bool = Depends(verify_battery_access),
):
    """
    Store the latest battery level for a device.

    Example: POST /api/battery?device=iphone&level=25
    """
    record = set_device_level(device, level)
    return BatteryUpdateResponse(success=True, **record)


@router.get("", response_model=Dict[str, BatteryLevel])
async def get_all_battery():
    """Get the latest battery level for all known devices. (Public)"""
    levels = get_all_levels()
    return {device: {"device": device, **record} for device, record in levels.items()}


@router.get("/{device}", response_model=BatteryLevel)
async def get_battery(device: str):
    """Get the latest battery level for a single device. (Public)"""
    record = get_device_level(device)
    if record is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No battery level recorded for device '{device}'",
        )
    return {"device": device, **record}

File: system/backend/src/routes/battery.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */



import { Router, type Request, type Response } from 'express';

import { getAllLevels, getDeviceLevel, setDeviceLevel } from '../services/battery_service.js';
import { verifyBatteryAccess } from '../dependencies/battery.js';
import { asString } from '../utils/request.js';

export const batteryRouter = Router();

/**
 * Store the latest battery level for a device.
 * Example: POST /api/battery?device=iphone&level=25
 */
batteryRouter.post('/api/battery', verifyBatteryAccess, async (req: Request, res: Response) => {
  const device = req.query.device;
  const levelRaw = req.query.level;

  if (typeof device !== 'string' || device.length < 1 || device.length > 64) {
    res.status(422).json({ detail: "Query param 'device' must be a string between 1 and 64 characters" });
    return;
  }

  const level = Number(levelRaw);
  if (typeof levelRaw !== 'string' || Number.isNaN(level) || level < 0 || level > 100) {
    res.status(422).json({ detail: "Query param 'level' must be an integer between 0 and 100" });
    return;
  }

  const record = await setDeviceLevel(device, level);
  res.json({ success: true, ...record });
});

/** Get the latest battery level for all known devices. (Public) */
batteryRouter.get('/api/battery', async (_req: Request, res: Response) => {
  const levels = await getAllLevels();
  const result: Record<string, unknown> = {};
  for (const [device, record] of Object.entries(levels)) {
    result[device] = { device, ...record };
  }
  res.json(result);
});

/** Get the latest battery level for a single device. (Public) */
batteryRouter.get('/api/battery/:device', async (req: Request, res: Response) => {
  const { device } = req.params;
  const record = await getDeviceLevel(asString(device));
  if (record === null) {
    res.status(404).json({ detail: `No battery level recorded for device '${device}'` });
    return;
  }
  res.json({ device, ...record });
});